editor/plugins/version_control_editor_plugin.cpp

```diff
@@ -240,42 +353,101 @@ void VersionControlEditorPlugin::_stage_selected() {
+	_refresh_branch_list();
+}
+
+void VersionControlEditorPlugin::_pull_pressed() {
+	EditorVCSInterface::get_singleton()->pull();
+	_refresh_stage_area();
+	_refresh_branch_list();
+	_refresh_commit_list();
+}
+
+void VersionControlEditorPlugin::_push_pressed() {
+	EditorVCSInterface::get_singleton()->push();
+}
+
+void VersionControlEditorPlugin::_move_all(Object *p_from, Object *p_to) {
 
 	if (!EditorVCSInterface::get_singleton()) {
 
 		WARN_PRINT("No VCS addon is initialized. Select a Version Control Addon from Project menu");
 		return;
 	}
 
-	staged_files_count = 0;
-	TreeItem *root = stage_files->get_root();
+	Tree *from = Object::cast_to<Tree>(p_from);
+	Tree *to = Object::cast_to<Tree>(p_to);
+
+	TreeItem *root = from->get_root();
 	if (root) {
 
 		TreeItem *file_entry = root->get_children();
 		while (file_entry) {
-
-			EditorVCSInterface::get_singleton()->stage_file(file_entry->get_metadata(0));
-			file_entry->set_icon_modulate(0, EditorNode::get_singleton()->get_gui_base()->get_color("success_color", "Editor"));
-			file_entry->set_checked(0, true);
-			staged_files_count++;
-
+			_move_item(from, to, file_entry);
 			file_entry = file_entry->get_next();
 		}
 	}
-
 	_update_stage_status();
+	_refresh_stage_area();
 }
 
-void VersionControlEditorPlugin::_view_file_diff() {
+void VersionControlEditorPlugin::_view_file_diff(Object *tree) {
 
 	version_control_dock_button->set_pressed(true);
 
-	String file_path = stage_files->get_selected()->get_metadata(0);
+	String file_path = Object::cast_to<Tree>(tree)->get_selected()->get_meta("file_path");
 
 	_display_file_diff(file_path);
 }
 
+void VersionControlEditorPlugin::_item_activated(Object *p_from, Object *p_to) {
+	Tree *from = Object::cast_to<Tree>(p_from);
+	Tree *to = Object::cast_to<Tree>(p_to);
+	TreeItem *item = from->get_selected();
+	_move_item(from, to, item);
+	_refresh_stage_area();
+}
+
+void VersionControlEditorPlugin::_move_item(Tree *from, Tree *to, TreeItem *item) {
+	if (from == staged_files) {
+		EditorVCSInterface::get_singleton()->unstage_file(item->get_meta("file_path"));
+	} else {
+		EditorVCSInterface::get_singleton()->stage_file(item->get_meta("file_path"));
+	}
+}
+
+void VersionControlEditorPlugin::_cell_button_pressed(Object *p_item, int column, int id) {
+	TreeItem *item = Object::cast_to<TreeItem>(p_item);
+	String file_path = item->get_meta("file_path");
+	ChangeType change = (ChangeType)(int)item->get_meta("change_type");
+
+	if (id == BUTTON_TYPE_OPEN && change != CHANGE_TYPE_DELETED) {
+
+		Error err;
+		RES res = ResourceLoader::load(file_path, "", false, &err);
+		if (!res.is_valid()) {
+			return;
+		}
+
+		List<String> scene_extensions;
+		ResourceLoader::get_recognized_extensions_for_type("PackedScene", &scene_extensions);
+
+		if (scene_extensions.find(file_path.get_extension())) {
+			//FIXIT: Alway opens new tab regardless of scene is open or not.
+			EditorNode::get_singleton()->load_scene(file_path);
```

Comment:
I am not sure why this is needed. I will build and run and check this again

Reply:
The stage and unstage have a button to open the changed file. It opens the same scene again and again in new tabs.